certbot/util.py

```diff
@@ -66,28 +66,36 @@ def run_script(params):
     return stdout, stderr
 
 
-def exe_exists(exe):
+def _is_exe(fpath):
+    return os.path.isfile(fpath) and os.access(fpath, os.X_OK)
+
+def which(program):
     """Determine whether path/name refers to an executable.
 
-    :param str exe: Executable path or name
+    :param str program: Executable path or name
 
-    :returns: If exe is a valid executable
-    :rtype: bool
+    :returns: Path to executable if it exists, or None
+    :rtype: str or None
 
     """
-    def is_exe(path):
-        """Determine if path is an exe."""
-        return os.path.isfile(path) and os.access(path, os.X_OK)
-
-    path, _ = os.path.split(exe)
-    if path:
-        return is_exe(exe)
+    # Borrowed from:
+    # https://stackoverflow.com/questions/377017/test-if-executable-exists-in-python
+    # XXX May need more porting to handle .exe extensions on Windows
+
+    fpath, _fname = os.path.split(program)
+    if fpath:
+        if _is_exe(program):
+            return program
     else:
         for path in os.environ["PATH"].split(os.pathsep):
-            if is_exe(os.path.join(path, exe)):
-                return True
+            exe_file = os.path.join(path, program)
+            if _is_exe(exe_file):
+                return exe_file
+
+    return None
 
-    return False
+# Either name makes sense, and we have a lot of call sites!
+exe_exists = which
```

Comment:
What about simply using the existing `exe_exists` in your code? There are two call sites to `which` with one `mock.patch` with 21 calls/mocks for `exe_exists`. Why have two functions with the same name? You also only ever use the result of `which` as a `bool` except when you call `_prog` [here](https://github.com/certbot/certbot/pull/3918/files#diff-fee68073ff44ffe30e9303db0210c070R115).

I think this call should be changed. If `which` fails due to known problems with hook validation, the user will get a message like "Error output from None". If `which` succeeds, you pass it in to `os.path.basename` which I believe will return the same thing whether or not you called `which` (`os.path.basename("/bin/echo") == os.path.basename("echo")`).